backend/app/libs/tokenlists.py

```diff
@@ -0,0 +1,71 @@
+from asyncio import gather, sleep
+from json.decoder import JSONDecodeError
+from typing import Coroutine
+
+from celery.utils.log import get_task_logger
+from httpx import AsyncClient, HTTPStatusError, RequestError, Timeout
+
+from .storage_helpers import retrieve_token_whitelist, store_token_whitelist
+
+
+async def retry(method: Coroutine, url: str, await_time: float):
+    await sleep(await_time)
+    await method(url)
+
+
+async def get_token_list(token_list: str):
```

Comment:
I would recommend to try keeping functions even simpler. Here is a pattern which I believe is going to help keep functions as simple as possible and help you write tests faster:

- `get_token_list`: performs some HTTP GET, returns the decoded JSON, and raises `RequestError`, `HTTPStatusError`, `KeyError`, `JSONDecodeError`.
- `maybe_populate_whitelist`: manages retrieving as well as **storing** in db, calls the `get_token_list` function and **handles raised errors itself**.

This way, in your tests, you can write much simpler tests for `get_token_list`, for each test case:

- Mock `AyncClient`, `raise_for_status`…
- Check that the right error is raised.

(gotcha: these are not the most useful tests, as we are in a large extent just testing the underlying asyncio functions, but that's still useful to **document the exceptions raised** in a robust way)

Now, for `maybe_populate_whitelist`, you can:

- Test the log messages. But to be honest, I think this is going to be a pretty useless test for now (we could decide not log anything, that's not a core feature of the application itself).
- Test the fallback behavior in case of any kind of error from `get_token_list`.
- Test the good scenario: if `get_token_list` returns something, it should be stored in db and passed along as a return value.

Reply:
`get_all_token_lists` also depends on `get_token_list` and is also called independently to `maybe_populate_whitelist`.

See [here](https://github.com/butterdao/whip/pull/52/files#diff-d2c8335ca862b4620540396ce17363eefb2870e847c1fcb8f729640f895e9109R410). Which means only doing the error handling in one or the other would leave the other neglected. I can just do away with this [independent call](https://github.com/butterdao/whip/pull/52/files#diff-d2c8335ca862b4620540396ce17363eefb2870e847c1fcb8f729640f895e9109R410) if necessary, but this check would render this periodic task
```python
sender.add_periodic_task(
        crontab(hour=0, minute=0, day_of_week=[1], nowfun=datetime.now),
        reload_whitelist.s(),
        name="reload token whitelist",
    )
```
and [calls to reload the whitelist on program initialization](https://github.com/butterdao/whip/pull/52/files#diff-d2c8335ca862b4620540396ce17363eefb2870e847c1fcb8f729640f895e9109R349-R351) useless if Heroku keeps the local cache persistent between deployments/restarts.

Or I can take your suggestion and let the independent call fail without fullback. If we don't add any retry logic to response errors it can leave the token list outdated until next scheduled execution (but then again if the retry fails, same outcome).

I will take the rest of your suggestions into account 🙂.

Reply:
As discussed, better structure has been added to `tokenlists` and tests have been modified to comply with changes.